### PR TITLE
Adding full finetuning

### PR DESCRIPTION
**Pull request by Goekdeniz-Guelmez**
As before, this pull request contains changes to the code to feature full finetuning, files changed are lora.py, trainer,py and the LORA.md file for the new arguments.

the new training arguments are now:
```
python -m mlx_lm.lora \
    --model  \
    --train \
    --fine-tune-type full \
    --data  \
    --iters 100 \
    --batch-size 1 \
    --val-batches 1 \
    --adapter-path 
```

to change the method of finetuning, simply change `--fine-tune-type` to `dora`, `lora`, or `full`, default is LoRA. Path where the adapters of the full model weights are stored is still `--adapter-file`.

### CODE REVIEW

**Comment by Goekdeniz-Guelmez** [comment]
Tested with gemma, mistral, llama3 (tiny versions because i only have a 8GB Macbook Air):


FULL
```
python -m mlx_lm.lora \
    --model /Users/gokdenizgulmez/Library/Mobile\ Documents/com\~apple\~CloudDocs/Transformer\ Models/Safetensors/tiny-random-GemmaForCausalLM \
    --train \
    --fine-tune-type full \
    --data /Users/gokdenizgulmez/Library/Mobile\ Documents/com\~apple\~CloudDocs/Datastes/data_tyni \
    --iters 10 \
    --batch-size 1 \
    --val-batches 1 \
    --adapter-path /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain
Loading pretrained model
Loading datasets
Training
Training full model weights.
Trainable parameters: 100.000% (2.049M/2.049M)
Starting training..., iters: 10
Iter 1: Val loss 12.459, Val took 0.455s
Iter 2: Saved model checkpoint weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain/model.safetensors and /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain/0000002_checkpoint.safetensors.
Iter 4: Saved model checkpoint weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain/model.safetensors and /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain/0000004_checkpoint.safetensors.
Iter 6: Saved model checkpoint weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain/model.safetensors and /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain/0000006_checkpoint.safetensors.
Iter 8: Saved model checkpoint weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain/model.safetensors and /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain/0000008_checkpoint.safetensors.
Iter 10: Val loss 12.456, Val took 0.346s
Iter 10: Train loss 12.456, Learning Rate 1.000e-05, It/sec 17.743, Tokens/sec 17372.541, Trained Tokens 9791, Peak mem 6.281 GB
Iter 10: Saved model checkpoint weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain/model.safetensors and /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain/0000010_checkpoint.safetensors.
Saved final full model weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain/model.safetensors.
```

LORA
```
python -m mlx_lm.lora \
    --model /Users/gokdenizgulmez/Library/Mobile\ Documents/com\~apple\~CloudDocs/Transformer\ Models/Safetensors/tiny-random-GemmaForCausalLM \
    --train \
    --fine-tune-type lora \
    --data /Users/gokdenizgulmez/Library/Mobile\ Documents/com\~apple\~CloudDocs/Datastes/data_tyni \
    --iters 10 \
    --batch-size 1 \
    --val-batches 1 \
    --adapter-path /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-lora
Loading pretrained model
Loading datasets
Training
Training model with LoRA.
Trainable parameters: 0.022% (0.000M/2.049M)
Starting training..., iters: 10
Iter 1: Val loss 12.459, Val took 0.372s
Iter 2: Saved adapter weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-lora/adapter.safetensors and /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-lora/0000002_adapters.safetensors.
Iter 4: Saved adapter weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-lora/adapter.safetensors and /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-lora/0000004_adapters.safetensors.
Iter 6: Saved adapter weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-lora/adapter.safetensors and /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-lora/0000006_adapters.safetensors.
Iter 8: Saved adapter weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-lora/adapter.safetensors and /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-lora/0000008_adapters.safetensors.
Iter 10: Val loss 12.457, Val took 0.219s
Iter 10: Train loss 12.457, Learning Rate 1.000e-05, It/sec 25.792, Tokens/sec 25253.294, Trained Tokens 9791, Peak mem 6.128 GB
Iter 10: Saved adapter weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-lora/adapter.safetensors and /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-lora/0000010_adapters.safetensors.
Saved final adapter weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-lora/adapter.safetensors.
```

DORA
```
python -m mlx_lm.lora \
    --model /Users/gokdenizgulmez/Library/Mobile\ Documents/com\~apple\~CloudDocs/Transformer\ Models/Safetensors/tiny-random-GemmaForCausalLM \
    --train \
    --fine-tune-type dora \
    --data /Users/gokdenizgulmez/Library/Mobile\ Documents/com\~apple\~CloudDocs/Datastes/data_tyni \
    --iters 10 \
    --batch-size 1 \
    --val-batches 1 \
    --adapter-path /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-dora
Loading pretrained model
Loading datasets
Training
Training model with DoRA.
Trainable parameters: 0.023% (0.000M/2.049M)
Starting training..., iters: 10
Iter 1: Val loss 12.460, Val took 0.241s
Iter 2: Saved adapter weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-dora/adapter.safetensors and /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-dora/0000002_adapters.safetensors.
Iter 4: Saved adapter weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-dora/adapter.safetensors and /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-dora/0000004_adapters.safetensors.
Iter 6: Saved adapter weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-dora/adapter.safetensors and /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-dora/0000006_adapters.safetensors.
Iter 8: Saved adapter weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-dora/adapter.safetensors and /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-dora/0000008_adapters.safetensors.
Iter 10: Val loss 12.457, Val took 0.164s
Iter 10: Train loss 12.457, Learning Rate 1.000e-05, It/sec 26.589, Tokens/sec 26033.065, Trained Tokens 9791, Peak mem 6.129 GB
Iter 10: Saved adapter weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-dora/adapter.safetensors and /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-dora/0000010_adapters.safetensors.
Saved final adapter weights to /Users/gokdenizgulmez/Desktop/tinyGemma-pretrain-dora/adapter.safetensors.
```

**Comment by awni** [comment]
Thanks!! Will review shortly!

**Comment by Goekdeniz-Guelmez** [comment]
> Thanks!! Will review shortly!

Perfect!

**Comment by Jonathan-Dobson** [comment]
Any idea how to correct this error?

```shell
  File ".../venv/lib/python3.8/site-packages/mlx/nn/utils.py", line 34, in wrapped_value_grad_fn
    value, grad = value_grad_fn(model.trainable_parameters(), *args, **kwargs)
RuntimeError: QuantizedMatmul::vjp no gradient wrt the quantized matrix yet.
```

**Comment by awni** [comment]
You can't fine-tune the quantized layers. You can use a fp16, bf16, or fp32 model for full fine-tuning. The half precision types need care to avoid numerical issues, so ymmv. If you want to use a quantized model you can do QLoRA.

**Comment by Jonathan-Dobson** [comment]
> You can't fine-tune the quantized layers. You can use a fp16, bf16, or fp32 model for full fine-tuning. The half precision types need care to avoid numerical issues, so ymmv. If you want to use a quantized model you can do QLoRA.

OK. Thanks for the info.
I probably won't attempt this on my 2021 M1
I was trying to fine tune  `mlx-community/DeepSeek-V2-Lite-Chat-4bit-mlx`
but failed with this error:
```shell
File ".../llms/mlx_lm/tuner/utils.py", line 132, in linear_to_lora_layers
    raise ValueError(f"Lora does not support {model.model_type}")
ValueError: Lora does not support deepseek_v2
```

**Comment by awni** [comment]
@Jonathan-Dobson here is a fix for that https://github.com/ml-explore/mlx-examples/pull/932. Will put it in a new pypi release once it lands.

**Comment by Jonathan-Dobson** [comment]
> @Jonathan-Dobson here is a fix for that #932. Will put it in a new pypi release once it lands.

 #932 fixed the error and allows Fine Tuning to start now.

**Comment by Jonathan-Dobson** [comment]
### Given a `--fine-tune-type full` training,

### When attempting to use generate.py like this:
```shell
  python -m mlx_lm.generate \
  --model mlx-community/Qwen2-0.5B \
  --prompt $P \
  --adapter-path adapters
```

### The command fails with this Error:
```shell
  File ".../venv/lib/python3.8/site-packages/mlx/nn/layers/base.py", line 204, in load_weights
    weights = list(mx.load(weights).items())
RuntimeError: [load_safetensors] Failed to open file adapters/adapters.safetensors
```

### Here are the contents of adapters after the full fine tune with updated lora.py:
```shell
adapter_config.json
model.safetensors
```

It looks like the full type saves the name as model now, but the generator.py still depends on the original adapters name.

**Comment by Goekdeniz-Guelmez** [comment]
It’s not saving all the model files correctly, what I tried is dragging the original config and tokenizer files to this `adapters` folder and then generating without the `—adapter` flag (and adding the adapters path to the`—model` flag), because it’s saving the full model weights, this flag is only needed for LoRA fine tuning. I’ll push the fix later, thanks for the feedback. Try it again after renaming the `adapter_config.json` to `config.json`.

**Comment by Goekdeniz-Guelmez** [comment]
Hey I'm back and it's fixed.

It now saves the full model with it's needed files like config.json, tokenizer, and so on.

You can now just generate with:

```sh
  python -m mlx_lm.generate \
  --model path/to/adapters \
  --prompt $P
```

**Comment by Goekdeniz-Guelmez** [comment]
Hey @awni, I want to ask if I need to do or change something for it to be merged?

**Comment by awni** [comment]
> Hey @awni, I want to ask if I need to do or change something for it to be merged?

Apologies for the delay. Let me take a look this week and get back to you!

File: llms/mlx_lm/tuner/utils.py [hunk 1]
```diff
@@ -14,6 +14,9 @@
 from .lora import LoRAEmbedding, LoRALinear, LoRASwitchLinear
 
 
+MAX_FILE_SIZE_GB = 5
```
Comment (awni) [inline llms/mlx_lm/tuner/utils.py:17]: unused?

File: llms/mlx_lm/tuner/utils.py [hunk 2]
```diff
@@ -2,7 +2,7 @@
 import json
 import types
 from pathlib import Path
-from typing import Dict
+from typing import Dict, Union, Any
```
Comment (awni) [inline llms/mlx_lm/tuner/utils.py:5]: Do you need those?

File: llms/mlx_lm/lora.py [hunk 3]
```diff
@@ -148,37 +155,48 @@ def build_parser():
...
+            print(f"Loading pretrained adapters from {args.resume_adapter_file}")
+            model.load_weights(args.resume_adapter_file, strict=False)
+    elif args.fine_tune_type == "dora":
+        print("Training model with DoRA.")
+        model.freeze()
+        # Convert linear layers to lora layers and unfreeze in the process
+        linear_to_lora_layers(model, args.lora_layers, args.lora_parameters, use_dora=True)
+        if args.resume_adapter_file is not None:
+            print(f"Loading pretrained adapters from {args.resume_adapter_file}")
+            model.load_weights(args.resume_adapter_file, strict=False, use_dora=True)
```
Comment (awni) [inline llms/mlx_lm/lora.py:188]: merge the lora and dora cases here to reduce half this code.

File: llms/mlx_lm/lora.py [hunk 4]
```diff
@@ -234,6 +254,13 @@ def run(args, training_callback: TrainingCallback = None):
     print("Loading pretrained model")
     model, tokenizer = load(args.model)
 
+    # # Load the full model state if available
+    # if args.fine_tune_type != "full" and args.adapter_path:
+    #     model_file = Path(args.adapter_path) / "model.safetensors"
+    #     if model_file.exists():
+    #         print(f"Loading full model state from {model_file}")
+    #         model.load_weights(str(args.adapter_path) / "model.safetensors", strict=True)
```
Comment (awni) [inline llms/mlx_lm/lora.py:262]: ?

**Comment by Goekdeniz-Guelmez** [comment]
Thanks for the quick reply! removed unnecessary code, merged the lora and dora cases and added the new lora layers.

File: llms/mlx_lm/tuner/utils.py (hunk 2 above)
Comment (Goekdeniz-Guelmez) [reply llms/mlx_lm/tuner/utils.py]: No, removed them.

**Review by awni — APPROVED** [review]
Thanks for the addition!!

**Comment by AtakanTekparmak** [comment]
@awni @Goekdeniz-Guelmez it could be nice to add a `FINETUNE.md` to [mlx-examples/llms/mlx_lm](https://github.com/ml-explore/mlx-examples/tree/main/llms/mlx_lm) imho

**Comment by Goekdeniz-Guelmez** [comment]
> @awni @Goekdeniz-Guelmez it could be nice to add a `FINETUNE.md` to [mlx-examples/llms/mlx_lm](https://github.com/ml-explore/mlx-examples/tree/main/llms/mlx_lm) imho

A nice and detailed description on how to fine-tunine is already in the LORA.md file [here](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/LORA.md#Run). Do you mean there shoud be a seperate and more detailed explaination?

**Comment by AtakanTekparmak** [comment]
> > @awni @Goekdeniz-Guelmez it could be nice to add a `FINETUNE.md` to [mlx-examples/llms/mlx_lm](https://github.com/ml-explore/mlx-examples/tree/main/llms/mlx_lm) imho
> 
> A nice and detailed description on how to fine-tunine is already in the LORA.md file [here](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/LORA.md#Run). Do you mean there shoud be a seperate and more detailed explaination?

Yeah I meant a separate file, I think full fine-tuning warrants its own guide as it's a feature long wanted for MLX. Great work btw 👏

**Pull request merged**